cgal@4.13 %oneapi@2023.2.1 fails build: Interval_nt.h: no member named 'get_static_minimin' in namespace 'CGAL::internal'

### Steps to reproduce the issue

`cgal@4.13 %oneapi@2023.2.1` fails to build using:
* spack@develop (e381e166ec27bb56c2d3bc856e7020224a005d1d from `Wed Aug 16 13:52:01 2023 -0700`)
* Ubuntu 20.04, amd64
* OneAPI 2023.2.1
* Docker container image `ecpe4s/ubuntu20.04-runner-amd64-oneapi-2023.2.1:2023.08.01`

<details><summary>Spack environment</summary>

```
spack:
  view: false

  concretizer:
    reuse: false
    unify: false

  compilers:
  - compiler:
      spec: oneapi@2023.2.1
      paths:
        cc: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/icx
        cxx: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/icpx
        f77: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/ifx
        fc: /opt/intel/oneapi/compiler/2023.2.1/linux/bin/ifx
      flags: {}
      operating_system: ubuntu20.04
      target: x86_64
      modules: []
      environment: {}
      extra_rpaths: []
  - compiler:
      spec: gcc@=11.4.0
      paths:
        cc: /usr/bin/gcc
        cxx: /usr/bin/g++
        f77: /usr/bin/gfortran
        fc: /usr/bin/gfortran
      flags: {}
      operating_system: ubuntu20.04
      target: x86_64
      modules: []
      environment: {}
      extra_rpaths: []

  packages:
    all:
      require: '%oneapi'
      providers:
        blas: [openblas]
        mpi: [mpich]
        tbb: [intel-tbb]
      target: [x86_64]
      variants: +mpi
    elfutils:
      variants: +bzip2 ~nls +xz
    hdf5:
      require: "%gcc"
      variants: +fortran +hl +shared
    libfabric:
      variants: fabrics=sockets,tcp,udp,rxm
    libunwind:
      variants: +pic +xz
    ncurses:
      variants: +termlib
    openblas:
      variants: threads=openmp
    trilinos:
      variants: +amesos +amesos2 +anasazi +aztec +belos +boost +epetra +epetraext
        +ifpack +ifpack2 +intrepid +intrepid2 +isorropia +kokkos +ml +minitensor +muelu
        +nox +piro +phalanx +rol +rythmos +sacado +stk +shards +shylu +stokhos +stratimikos
        +teko +tempus +tpetra +trilinoscouplings +zoltan +zoltan2 +superlu-dist gotype=long_long
    xz:
      variants: +pic
    mesa:
      version: [21.3.8]
    mpi:
      require: mpich
    mpich:
      require: '@4.1.2 ~wrapperrpath ~hwloc'
    py-cryptography:
      require: '@38.0.1'
    unzip:
      require: '%gcc'
    binutils:
      require: '%gcc'
      variants: +ld +gold +headers +libiberty ~nls
    llvm:
      require: '%gcc'
    ruby:
      require: '%gcc'
    rust:
      require: '%gcc'
    krb5:
      require: '%gcc'
    papi:
      require: '%gcc'
    openssh:
      require: '%gcc'
    libffi:
      require: "@3.4.4"
    dyninst:
      require: "%gcc"
    bison:
      require: '%gcc'

  specs:
  - openfoam
```

</details>

<details><summary>Concretization</summary>

<pre>
 -   atsrslh  cgal@4.13%oneapi@2023.2.1~core~demos+eigen~header_only~imageio~ipo+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64
[+]  f4pzjv5      ^boost@1.83.0%oneapi@2023.2.1+atomic+chrono~clanglibcpp~container~context~contract~coroutine+date_time~debug+exception~fiber+filesystem+graph~graph_parallel~icu+iostreams~json+locale+log+math+mpi+multithreaded~nowide~numpy~pic+program_options~python+random+regex+serialization+shared+signals~singlethreaded~stacktrace+system~taggedlayout+test+thread+timer~type_erasure~versionedlayout+wave build_system=generic cxxstd=98 patches=8e3faa2,a440f96,b8569d7 visibility=hidden arch=linux-ubuntu20.04-x86_64
[+]  vquxf4d          ^bzip2@1.0.8%oneapi@2023.2.1~debug~pic+shared build_system=generic arch=linux-ubuntu20.04-x86_64
[+]  nskvm3s              ^diffutils@3.9%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  nl5e3nn                  ^libiconv@1.17%oneapi@2023.2.1 build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  ifzxvxd          ^mpich@4.1.2%oneapi@2023.2.1~argobots~cuda+fortran~hwloc+hydra+libxml2+pci~rocm+romio~slurm~vci~verbs~wrapperrpath build_system=autotools datatype-engine=auto device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu20.04-x86_64
[+]  zs2iu4x              ^findutils@4.9.0%oneapi@2023.2.1 build_system=autotools patches=440b954 arch=linux-ubuntu20.04-x86_64
[+]  liacdsk              ^libfabric@1.18.1%oneapi@2023.2.1~debug~kdreg build_system=autotools fabrics=rxm,sockets,tcp,udp arch=linux-ubuntu20.04-x86_64
[+]  p4neb4n              ^libpciaccess@0.17%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  kktr6po                  ^util-macros@1.19.3%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  om2ykff              ^libxml2@2.10.3%oneapi@2023.2.1~python build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  2uey6zy              ^pkgconf@1.9.5%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  yjsjt4p              ^yaksa@0.2%oneapi@2023.2.1~cuda~rocm build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  jnrudht                  ^python@3.8.17%oneapi@2023.2.1+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,4c24573,f2fd060 arch=linux-ubuntu20.04-x86_64
[+]  kl2jtty                      ^expat@2.5.0%oneapi@2023.2.1+libbsd build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  rxlxaze                          ^libbsd@0.11.7%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  hfduf6w                              ^libmd@1.0.4%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  pnmat7v                      ^libffi@3.4.4%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  cixy4yv                      ^libxcrypt@4.4.35%oneapi@2023.2.1~obsolete_api build_system=autotools patches=4885da3 arch=linux-ubuntu20.04-x86_64
[+]  p4vso3z                      ^readline@8.2%oneapi@2023.2.1 build_system=autotools patches=bbf97f1 arch=linux-ubuntu20.04-x86_64
[+]  46bgn3p                      ^sqlite@3.42.0%oneapi@2023.2.1+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  mhquokx                      ^util-linux-uuid@2.38.1%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  l5aca6g          ^xz@5.4.1%oneapi@2023.2.1+pic build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  rjfoczb          ^zstd@1.5.5%oneapi@2023.2.1+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  bdi3o3i      ^cmake@3.26.3%oneapi@2023.2.1~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu20.04-x86_64
[+]  5udsvjx          ^curl@8.1.2%oneapi@2023.2.1~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu20.04-x86_64
[+]  yxh4tjy              ^nghttp2@1.52.0%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  6p2rujc              ^openssl@3.1.2%oneapi@2023.2.1~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu20.04-x86_64
[+]  r6unmd2                  ^ca-certificates-mozilla@2023-05-30%oneapi@2023.2.1 build_system=generic arch=linux-ubuntu20.04-x86_64
[+]  b4ekkqf          ^ncurses@6.4%oneapi@2023.2.1~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  vm2vkal      ^eigen@3.4.0%oneapi@2023.2.1~ipo build_system=cmake build_type=RelWithDebInfo generator=make arch=linux-ubuntu20.04-x86_64
[+]  vhkhu6i      ^gmake@4.4.1%oneapi@2023.2.1~guile build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  prscivl      ^gmp@6.2.1%oneapi@2023.2.1+cxx build_system=autotools libs=shared,static patches=69ad2e2 arch=linux-ubuntu20.04-x86_64
[+]  ztftzry          ^autoconf@2.69%oneapi@2023.2.1 build_system=autotools patches=35c4492,7793209,a49dd5b arch=linux-ubuntu20.04-x86_64
[+]  qymvhrq              ^perl@5.38.0%oneapi@2023.2.1+cpanm+opcode+open+shared+threads build_system=generic arch=linux-ubuntu20.04-x86_64
[+]  paauyc6                  ^berkeley-db@18.1.40%oneapi@2023.2.1+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu20.04-x86_64
[+]  475nhb5                  ^gdbm@1.23%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  brjlk4y          ^automake@1.16.5%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  yahjzmw          ^libtool@2.4.7%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  f7d2wvj          ^m4@1.4.19%oneapi@2023.2.1+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-x86_64
[+]  luam5ai              ^libsigsegv@2.14%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  dvbdxtv      ^mpfr@4.2.0%oneapi@2023.2.1 build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]  6p2by5p          ^autoconf-archive@2023.02.20%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  47tuqyi          ^texinfo@7.0.3%oneapi@2023.2.1 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  2vwor2h              ^gettext@0.21.1%oneapi@2023.2.1+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]  tzoycsx                  ^tar@1.34%oneapi@2023.2.1 build_system=autotools zip=pigz arch=linux-ubuntu20.04-x86_64
[+]  4qnnnej                      ^pigz@2.7%oneapi@2023.2.1 build_system=makefile arch=linux-ubuntu20.04-x86_64
[+]  7oozhgt      ^zlib@1.2.13%oneapi@2023.2.1+optimize+pic+shared build_system=makefile arch=linux-ubuntu20.04-x86_64
</pre></details>

Error:
```
$> spack -e . install /atsr
...
==> Installing cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea [21/21]
==> No binary for cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/3e/3e3dd7a64febda58be54c3cbeba329ab6a73b72d4d7647ba4931ecd1fad0e3bc.tar.xz
==> No patches needed for cgal
==> cgal: Executing phase: 'cmake'
==> cgal: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    '/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2023.2.1/gmake-4.4.1-vhkhu6iey6d6lg4dsq5jk47yxb5cgyan/bin/make' '-j16'

6 errors found in build log:
     133    In file included from /tmp/root/spack-stage/spack-stage-cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea/spack-src/include/CGAL/Bbox_2_Line_2_intersection_impl.h:32:
     134    In file included from /tmp/root/spack-stage/spack-stage-cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea/spack-src/include/CGAL/Simple_cartesian.h:29:
     135    In file included from /tmp/root/spack-stage/spack-stage-cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea/spack-src/include/CGAL/Cartesian/Cartesian_base.h:29:
     136    In file included from /tmp/root/spack-stage/spack-stage-cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea/spack-src/include/CGAL/basic.h:45:
     137    In file included from /tmp/root/spack-stage/spack-stage-cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea/spack-src/include/CGAL/number_type_basic.h:61:
     138    In file included from /tmp/root/spack-stage/spack-stage-cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea/spack-src/include/CGAL/long_double.h:38:
  >> 139    /tmp/root/spack-stage/spack-stage-cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea/spack-src/include/CGAL/Interval_nt.h:186:16: error: no member named 'get_static_minimin' in na
            mespace 'CGAL::internal'; did you mean simply 'get_static_minimin'?
     140        return IA(-CGAL_IA_MIN_DOUBLE, CGAL_IA_MIN_DOUBLE);
     141                   ^~~~~~~~~~~~~~~~~~
     142    /tmp/root/spack-stage/spack-stage-cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea/spack-src/include/CGAL/FPU.h:71:33: note: expanded from macro 'CGAL_IA_MIN_DOUBLE'
     143    #    define CGAL_IA_MIN_DOUBLE (CGAL::internal::get_static_minimin())
     144                                    ^~~~~~~~~~~~~~~~
     145    /tmp/root/spack-stage/spack-stage-cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea/spack-src/include/CGAL/FPU.h:121:9: note: 'get_static_minimin' declared here
...
  >> 173    /tmp/root/spack-stage/spack-stage-cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea/spack-src/include/CGAL/Interval_nt.h:738:36: error: no member named 'get_static_minimin' in na
            mespace 'CGAL::internal'; did you mean simply 'get_static_minimin'?
     174                          scale == 0 ? CGAL_IA_MIN_DOUBLE : scale);
     175                                       ^~~~~~~~~~~~~~~~~~
     176    /tmp/root/spack-stage/spack-stage-cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea/spack-src/include/CGAL/FPU.h:71:33: note: expanded from macro 'CGAL_IA_MIN_DOUBLE'
     177    #    define CGAL_IA_MIN_DOUBLE (CGAL::internal::get_static_minimin())
     178                                    ^~~~~~~~~~~~~~~~
     179    /tmp/root/spack-stage/spack-stage-cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea/spack-src/include/CGAL/FPU.h:121:9: note: 'get_static_minimin' declared here
     180    double& get_static_minimin(); // Defined in Interval_arithmetic_impl.h
     181            ^
     182    3 errors generated.
  >> 183    make[2]: *** [src/CGAL/CMakeFiles/CGAL.dir/build.make:79: src/CGAL/CMakeFiles/CGAL.dir/all_files.cpp.o] Error 1
     184    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea/spack-build-atsrslh'
  >> 185    make[1]: *** [CMakeFiles/Makefile2:197: src/CGAL/CMakeFiles/CGAL.dir/all] Error 2
     186    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-cgal-4.13-atsrslhoi7wsdsi7uyf75ueji3aaefea/spack-build-atsrslh'
  >> 187    make: *** [Makefile:139: all] Error 2
```


### Error message

<details><summary>Error message</summary>

<pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.21.0.dev0 (e381e166ec27bb56c2d3bc856e7020224a005d1d)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-skylake_avx512
* **Concretizer:** clingo

### Additional information

[CMakeOutput.log](https://github.com/spack/spack/files/12364344/CMakeOutput.log)
[spack-build-env.txt](https://github.com/spack/spack/files/12364345/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/12364346/spack-build-out.txt)


### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml
+++ b/share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml
@@ -124,6 +124,7 @@ spack:
   - netlib-scalapack
   - nrm
   - omega-h
+  - openfoam
   - openmpi
   - papi
   - papyrus
@@ -189,7 +190,6 @@ spack:
   # - glvis ^llvm                                             # glvis: https://github.com/spack/spack/issues/42839
   # - hpctoolkit                                              # dyninst@12.3.0%gcc: /usr/bin/ld: libiberty/./d-demangle.c:142: undefined reference to `_intel_fast_memcpy'; can't mix intel-tbb@%oneapi with dyninst%gcc
   # - mgard +serial +openmp +timing +unstructured ~cuda       # mgard: mgard.tpp:63:48: error: non-constant-expression cannot be narrowed from type 'int' to 'unsigned long' in initializer list [-Wc++11-narrowing]
-  # - openfoam                                                # cgal: https://github.com/spack/spack/issues/39481
   # - openpmd-api                                             # mgard:  mgard.tpp:63:48: error: non-constant-expression cannot be narrowed from type 'int' to 'unsigned long' in initializer list [-Wc++11-narrowing]
   # - pdt                                                     # pdt: pdbType.cc:193:21: warning: ISO C++11 does not allow conversion from string literal to 'char *' [-Wwritable-strings]
   # - quantum-espresso                                        # quantum-espresso@7.2 /i3fqdx5: warning: <unknown>:0:0: loop not unroll-and-jammed: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering

--- a/var/spack/repos/builtin/packages/openfoam/package.py
+++ b/var/spack/repos/builtin/packages/openfoam/package.py
@@ -369,8 +369,9 @@ class Openfoam(Package):
     # See https://github.com/spack/spack/pull/22303 for reference
     depends_on(Boost.with_default_variants)
 
-    # OpenFOAM does not play nice with CGAL 5.X
-    depends_on("cgal@:4")
+    # Earlier versions of OpenFOAM did not play nice with CGAL 5.X
+    depends_on("cgal")
+
     # The flex restriction is ONLY to deal with a spec resolution clash
     # introduced by the restriction within scotch!
     depends_on("flex@:2.6.1,2.6.4:")

--- a/var/spack/repos/builtin/packages/openfoam/package.py
+++ b/var/spack/repos/builtin/packages/openfoam/package.py
@@ -369,8 +369,11 @@ class Openfoam(Package):
     # See https://github.com/spack/spack/pull/22303 for reference
     depends_on(Boost.with_default_variants)
 
-    # Earlier versions of OpenFOAM did not play nice with CGAL 5.X
-    depends_on("cgal")
+    # Earlier versions of OpenFOAM may not work with CGAL 5.6. I do
+    # not know which OpenFOAM added support for 5.x and conservatively
+    # use 2312 in the check.
+    depends_on("cgal", when="@2312:")
+    depends_on("cgal@:4", when="@:2306")
 
     # The flex restriction is ONLY to deal with a spec resolution clash
     # introduced by the restriction within scotch!

